### PR TITLE
Azdevify chat-copilot

### PR DESCRIPTION
**Pull request by Menghua1**
This pr is all changes to convert `chat-copilot` application into an Azd template.

You can currently deploy this application using azd by following the steps below:

1. Execute the command: `azd init -t Menghua1/chat-copilot -b azdevify`.
2. Execute the command: `azd up`.

The reasons for not using keyvault to get keys are as follows:

- `AzureOpenAIEmbedding Key`, `AzureCognitiveSearch Key`  and `AzureOpenAIText Key`: Because these three keys are stored in `KernelMemory`, dotnet core will read them through `AddUserSecrets` during `AddConfiguration` and then add them to `configBuilder`. The values for these keys must be provided in `KernelMemory`.

image
image

The reasons for using prepackage hooks in `azure.yaml`: When `azd deploy`, the front end uses npm to load dependencies, which may cause dependency incompatibility issues.

@jongio for notification.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,7 +32,46 @@ You will need the following items to run the sample:
...
+
+After logging in with the following command, you will be able to use the `azd` cli to quickly provision and deploy the application.
+
+```
+azd auth login
+```
+
+Then, execute the `azd init` command to initialize the environment (You do not need to run this command if you already have the code or have opened this in a Codespace or DevContainer).
+```
+azd init -t jongio/chat-copilot
```
Comment (jongio) [inline README.md:61]: Please update this to `microsoft/chat-copilot`

File: infra/app/api.bicep [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+param appInsightsConnectionString string
+param azureCognitiveSearchName string
+param deployWebSearcherPlugin bool
+param allowedOrigins array = []
+param functionAppWebSearcherPluginName string
+param searcherPluginDefaultHostName string
+param openAIServiceName string
+param openAIEndpoint string
+param storageAccountName string
+param virtualNetworkId0 string
```
Comment (jongio) [inline infra/app/api.bicep:19]: why '0' at end of name

File: infra/app/storage.bicep [hunk 3]
```diff
@@ -0,0 +1,25 @@
+param name string
```
Comment (jongio) [inline infra/app/storage.bicep:1]: why not infra/core?

File: infra/app/api.bicep [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+var functionAppWebSearcherPluginId = resourceId(subscription().subscriptionId, resourceGroup().name,
+  'Microsoft.Web/sites', functionAppWebSearcherPluginName)
+var openAIId = resourceId(subscription().subscriptionId, resourceGroup().name,
+  'Microsoft.CognitiveServices/accounts', openAIServiceName)
+var storageAccountId = resourceId(subscription().subscriptionId, resourceGroup().name,
+  'Microsoft.Storage/storageAccounts', storageAccountName)
+var speechAccountId = resourceId(subscription().subscriptionId, resourceGroup().name,
+  'Microsoft.CognitiveServices/accounts', speechAccountName)
+var cosmosAccountId = resourceId(subscription().subscriptionId, resourceGroup().name,
+  'Microsoft.DocumentDB/databaseAccounts', cosmosAccountName)
```
Comment (jongio) [inline infra/app/api.bicep:40]: What we typically do I pass the name and then get a reference to the storage account with 'existing' keyword in bicep and then use that, 

like

resource storage '' = existing {
   name:
}

then 

storage.listKeys

Comment (jongio) [reply infra/app/api.bicep:40]: Can you see that is possible?

**Review by jongio — CHANGES_REQUESTED** [review]

File: infra/app/storage.bicep (hunk 3 above)
Comment (Menghua1) [reply infra/app/storage.bicep]: We need to create `fileshares` with `StorageAccount` as the parent, but `core/storage/storage-account.bicep` does not support this part. Based on your [comment](https://github.com/jongio/chat-with-your-data-solution-accelerator/pull/2#discussion_r1449306153) in another sample, updated the `core/storage/storage-account.bicep` file to support this.

File: README.md (hunk 1 above)
Comment (Menghua1) [reply README.md]: Updated.

File: infra/app/api.bicep (hunk 2 above)
Comment (Menghua1) [reply infra/app/api.bicep]: Update to `virtualNetworkId`.

File: infra/app/api.bicep (hunk 4 above)
Comment (Menghua1) [reply infra/app/api.bicep:40]: It's possible, but based on your [comment](https://github.com/jongio/chat-with-your-data-solution-accelerator/pull/2#discussion_r1383732678) in another sample, the "existing" call is a costly ARM deployment, and there are five services that need to get keys by referencing "existing". Are you sure change these five services to reference the "existing" keyword?

Comment (jongio) [reply infra/app/api.bicep:40]: My point on that was that if we already have the reference then we shouldn't need to query it again.  Let's keep it as-is for now and not switch to existing.  Thanks

**Pull request merged**